### ISSUE TITLE
具体代码 preprocess_true_boxes（）：tw th 定义是什么? 

### ISSUE THREAD

**Issue by ningxiangyun** [raytroop/YOLOv3_tf#7]
idx = cxy[..., 1] * (num_anchors * feat_size[2]) + num_anchors * cxy[..., 0] + iout_argmax 
我看的有点懵逼。
大佬能给说下吗

**Comment by raytroop** [comment]
I think they are used to predict  width/height of bounding box.
You may read YOLOv3 paper for details.

![image](https://user-images.githubusercontent.com/31182666/41296530-b67eb1c2-6e8f-11e8-88ab-cecb14e095e6.png)

**Comment by stickOverCarrot** [comment]
@ningxiangyun 这个是在求解哪个位置的哪个预选框来负责探测标签box。首先得明白，该层feature_map有feat_size[...,1]*feat_size[...,2]*3的预选框。用坐标表示就是（y,x,n）如果reshape成1行，那么idex=y*feat_size[...,2]*3+x*3+n,自己画个图就明白了.
,也就是该coder为什么这样写的原因。
大致上看了该作者的思路，感觉计算loss上有错误，如果大家也有觉得有错误，可以跟我探讨

**Comment by lkj1114889770** [comment]
@stickOverCarrot 感觉作者的代码和yolo3官方的loss计算方法并不一致 
![qq 20180615140713](https://user-images.githubusercontent.com/19361626/41452648-6cbeb204-70a5-11e8-8ede-f4275f327e38.png)
根据这个公式坐标误差应该有开根号w h后相减平方和的部分，但是代码中用直接相减平方，

`matching_classes = tf.cast(matching_true_boxes[..., 4], tf.int32)   # [batch, 13, 13, 3]
    matching_classes = tf.one_hot(matching_classes, num_classes)    # [batch, 13, 13, 3, 20]
    classification_loss = (detectors_mask *
                           tf.square(matching_classes - pred_class_prob))   # [batch, 13, 13, 3, 20]

    # Coordinate loss for matching detection boxes.   [σ(tx), σ(ty), tw, th]
    matching_boxes = matching_true_boxes[..., 0:4]
    coordinates_loss = (detectors_mask * loc_scale * tf.square(matching_boxes - pred_boxes))

    classification_loss_sum = tf.reduce_sum(classification_loss)
    coordinates_loss_sum = tf.reduce_sum(coordinates_loss)`

诸如此类，前面的confidence_loss计算方式和原先YOLO论文中的方法也并不一样

**Comment by lkj1114889770** [comment]
@stickOverCarrot 用了一个不太一样的方法计算loss的感觉，所以感觉一开始都没看懂，比如这个coordinate_loss,前面还计算了个loc_scale来作为系数相乘的感觉，前面用2来相减得到

` loc_scale = 2 - true_boxes[..., 2] * true_boxes[..., 3]     # (batch, 30)
`
对于大框系数会变小，也起到了平衡大小框loss的效果，个人理解作者这个代码用了不一样的loss计算。如果看法有偏差的话，欢迎继续探讨。

**Comment by ningxiangyun** [comment]
你们对preprocess_true_boxes（）掩码有木有觉得奇怪?

**Comment by zhliu2018** [comment]
我也觉得loss计算有问题，掩码应该不对，大家可以加QQ:2901129562讨论一下，如果人多的话就拉个群 @ningxiangyun @lkj1114889770 @stickOverCarrot

**Comment by mrhalyang** [comment]
@lkj1114889770 loc_scale = 2 - true_boxes[..., 2] * true_boxes[..., 3] # (batch, 30) 的实现在Redmon的darknet代码中也有[yolo_layer.c](https://github.com/pjreddie/darknet/blob/98e2a7674fd6a81349735db6f5b34229c871113a/src/yolo_layer.c#L190)

**Comment by lkj1114889770** [comment]
@wearyoung 噢，看了一下确实是这样的，是类似的实现
